### ISSUE TITLE
How to change vocabulary size in transformer model? 

### ISSUE THREAD

**Issue by leixxli** [pytorch/fairseq#1909]
Hi, I am using Transformer model to translate English to Spanish. 

First, I trained transformer in Europarl data.  The vocabulary size is 7376 for english and 9216 for Spanish. Then, I finetuned a smaller dataset on this model, the vocabulary size is 6447 types for english and 8783 in Spanish.

The problem is, the vocabulary size didn't match, so I can't continue training extra steps in this smaller dataset, using my pretrained model. 

I have tried using fairseq-proprocess --tgtdict and --srcdict, to preprocess my smaller dataset based on Europarl vocabulary, which will give me the same dimension of vocabulary. However, words that in my smaller dataset has many terminologies, which does not contained in Europral vocabulary.  Using -tgtdict and --srcdict will sacrifice these words into unknown.  The bleu score is obviously poor. 

The script I am using is as follows: it haves the dimension mismatch problem. 
fairseq-train \
    data-bin/smaller_dataset \
    --restore-file checkpoints/Europarl/checkpoint_best.pt \             # vocabulary size: (7376, 9216)
    --arch transformer --dropout 0.1 \                                                  # vocabulary size:  (6447, 8783) 
    --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
    --lr 8e-4 --lr-scheduler inverse_sqrt --warmup-updates 4000 \
    --dropout 0.3 --weight-decay 0.0001 \
    --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
    --max-tokens 4096 \
    --fp16 \
    --save-dir checkpoints/result

thanks.

**Comment by villmow** [comment]
Use byte-pair-encoding before  preprocessing your dataset with fairseq (e.g. subword-nmt).

**Comment by erip** [comment]
You'll also want to maintain the source and target dicts from pretraining when preprocessing the fine-tuning dataset.

**Comment by leixxli** [comment]
Thanks for the reply. @villmow 

I can either change the BPE_TOKENS in subword-nmt stage, or preprocessing the combined dataset, then using this dictionary to generate individual dictionary for each dataset.
